### PR TITLE
Adding unit tests for EnsureVisibleDuration

### PR DESCRIPTION
**Pull request by amanda-tarafa**
And fixing and edge case.

### CODE REVIEW

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs [hunk 1]
```diff
@@ -79,6 +84,49 @@ private static bool Is1MsSpan(TraceSpan span)
...
         }
 
+        /// <summary>
+        /// Used to ensure that <see cref="Timestamp"/>s manipulated by
+        /// <see cref="SimpleManagedTracer"/> are normalized after
+        /// manipulation.
+        /// </summary>
+        private static bool IsNormalized(Timestamp timestamp) =>
+            timestamp.Nanos >= 0 &&
+            timestamp.Nanos <= MaxNanos &&
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs]: I think I'd just use Duration.NanosecondsPerSecond and make the range exclusive.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs [hunk 2]
```diff
@@ -79,6 +84,49 @@ private static bool Is1MsSpan(TraceSpan span)
...
+            timestamp.Seconds <= UnixSecondsAtBclMaxValue;
+
+        [Theory]
+        [InlineData(10, 10, 10, 15)]
+        [InlineData(10, 999000000, 10, 999000015)]
+        [InlineData(10, 999999999, 11, 15)]
+        [InlineData(10, 15, 10, 10)]
+        public void Span_EnsureVisibleDuration(int startSeconds, int startNanos, int endSeconds, int endNanos)
+        {
+            var mockConsumer = new Mock<IConsumer<TraceProto>>();
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs]: What's this mock actually doing? If we don't care about how it responds, I think I'd normally create a "no-op" fake instead of mocking.

**Comment by jskeet** [comment]
I think I'd probably change the existing calls to check that every call's span is *at least* 1ms instead. Deterministic, but still cautious.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs]: Done

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs]: Done

**Comment by amanda-tarafa** [comment]
Adressed PR comments, added test from when span is longer than 1ms and changed verification of span length to at least 1ms. PTAL.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        /// Does nothing
+        /// </summary>
+        /// <param name="items"></param>
+        public void Receive(IEnumerable<TraceProto> items) { }
+
+        /// <summary>
+        /// Returns a completed task.
+        /// </summary>
+        public Task ReceiveAsync(IEnumerable<TraceProto> items, CancellationToken cancellationToken = default) =>
+#if NET452
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs]: I'd just use `Task.FromResult(0)` in both cases, rather than make it conditional.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using TraceProto = Google.Cloud.Trace.V1.Trace;
+
+namespace Google.Cloud.Diagnostics.Common.Tests
+{
+    /// <summary>
+    /// An <see cref="IConsumer{T}"> for testing that does nothing.
+    /// </summary>
+    public class FakeConsumer : IConsumer<TraceProto>
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs]: *Possibly* rename to NoOpConsumer as later we may want a FakeConsumer that remembers all the traces it receives? Although it wouldn't hurt to add that functionality to this one instead...

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs [hunk 5]
```diff
@@ -73,10 +77,86 @@ private static bool IsValidSpan(TraceSpan span, string name, ulong parentId, Spa
         /// a difference of 1ms between start and end. If we do not the Trace
         /// API will not record the span.
         /// </summary>
-        private static bool Is1MsSpan(TraceSpan span)
+        private static bool IsAtLeast1MsSpan(TraceSpan span)
         {
             var duration = span.EndTime - span.StartTime;
-            return duration.Seconds == 0 && duration.Nanos == 1_000_000;
+            return duration.Seconds >= 1 || duration.Nanos >= 1_000_000;
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs]: To be absolutely sure, probably

```csharp
duration.Seconds >= 1 || (duration.Seconds == 0 && duration.Nanos >= 1_000_000);
```

Otherwise if we had our start and end time the wrong way round we'd allow a negative value to pass.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs (hunk 3 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs]: Done, only I did Task.Delay(0) in both cases.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs (hunk 4 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs]: I'm renaming to NoOpConsumer. In real code there's no way for clients of IConsumer to access the traces it receives or any information about them, so I'm not sure why we would need a Fake to remember them? Unless we'd want to either verify something or expect some very specific behavior from the fake in which case we would use a Mock or write a specific Fake? Am I missing something?
If I'm wrong we can always add the remembering Fake when it's needed.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs (hunk 5 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs]: Yep, will change and add a comment or else that second check on seconds looks redundant.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs (hunk 4 above)
Comment (jskeet) [reply apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/FakeTracer.cs]: We'd want a fake to remember them so that we could check what was passed to the consumer. I typically prefer fakes over mocks as they're less brittle in the face of irrelevant changes, but it would depend on the specific situation. I'm fine with it being NoOpConsumer for now, certainly :)

**Comment by amanda-tarafa** [comment]
Thanks! Waiting on all green before merging.
